### PR TITLE
fix(Core/Spells): Luffa removing bleeds over level 60

### PR DESCRIPTION
**Pull request by avarishd**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes https://github.com/chromiecraft/chromiecraft/issues/3010
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/10711

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [x] Let me know if there is a better way to do this.
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/Spells/spell_item.cpp [hunk 1]
```diff
@@ -4092,6 +4092,62 @@ class spell_item_skyguard_blasting_charges : public SpellScript
...
+{
+    PrepareSpellScript(spell_item_luffa);
+    std::list<uint32> removeList;
+    std::list<ObjectGuid> casterList;
+
+    SpellCastResult CheckCast()
+    {
+        if (Unit* caster = GetCaster())
+        {
+            Unit::AuraApplicationMap const& auras = caster->GetAppliedAuras();
```
Comment (sogladev) [inline src/server/scripts/Spells/spell_item.cpp:4106]: no need to nullcheck `GetCaster()` .  I see `caster` is only used here, so you can use `GetCaster()` instead

File: src/server/scripts/Spells/spell_item.cpp [hunk 2]
```diff
@@ -4092,6 +4092,62 @@ class spell_item_skyguard_blasting_charges : public SpellScript
...
+            for (Unit::AuraApplicationMap::const_iterator itr = auras.begin(); itr != auras.end(); ++itr)
+            {
+                Aura const* aura = itr->second->GetBase();
+                if (aura->GetSpellInfo()->GetEffectMechanic(EFFECT_0) != MECHANIC_BLEED || aura->GetCasterLevel() > 60 || aura->GetSpellInfo()->IsPositive())
```
Comment (sogladev) [inline src/server/scripts/Spells/spell_item.cpp:4110]: `GetEffectMechanic(EFFECT_0)` only checks for `EFFECT_0`. You can check for all effects with `aura->GetSpellInfo()->GetAllEffectsMechanicMask() & (1 << MECHANIC_BLEED)`

I don't think saving to lists here is needed. `CheckCast()` is [used to override the usual result](https://github.com/avarishd/azerothcore-wotlk/blob/5a13809e8e9f491da95e13e47bab29bc718dd68d/src/server/game/Spells/Spell.cpp#L6072). returning `SPELL_FAILED_NOTHING_TO_DISPEL` here if we find a bleed by a `GetCasterLevel() > 60` is enough. The rest can be done in `HandleEffect`

File: src/server/scripts/Spells/spell_item.cpp [hunk 3]
```diff
@@ -4092,6 +4092,62 @@ class spell_item_skyguard_blasting_charges : public SpellScript
...
+
+        if (removeList.empty() || casterList.empty())
+            return SPELL_FAILED_NOTHING_TO_DISPEL;
+        else
+            return SPELL_CAST_OK;
+    }
+
+    void HandleEffect(SpellEffIndex effIndex)
+    {
+        PreventHitDefaultEffect(effIndex);
```
Comment (sogladev) [inline src/server/scripts/Spells/spell_item.cpp:4126]: this prevents the default [`void Spell::EffectDispelMechanic(SpellEffIndex effIndex)`](https://github.com/avarishd/azerothcore-wotlk/blob/5a13809e8e9f491da95e13e47bab29bc718dd68d/src/server/game/Spells/SpellEffects.cpp#L5131)

maybe copy the above and adjust for this specific case:

OwnedAuras -> `GetAppliedAuras()`
dispel_list -> no need to store. RemoveAura and return

**Comment by avarishd** [comment]
Much appreciated @sogladev!

File: src/server/scripts/Spells/spell_item.cpp [hunk 4]
```diff
@@ -4092,6 +4092,55 @@ class spell_item_skyguard_blasting_charges : public SpellScript
...
+        {
+            Unit::AuraApplicationMap const& auras = player->GetAppliedAuras();
+            for (Unit::AuraApplicationMap::const_iterator itr = auras.begin(); itr != auras.end(); ++itr)
+            {
+                Aura const* aura = itr->second->GetBase();
+                if (!(aura->GetSpellInfo()->GetAllEffectsMechanicMask() & (1 << MECHANIC_BLEED)) || aura->GetCasterLevel() > 60 || aura->GetSpellInfo()->IsPositive())
+                    continue;
+
+                    player->RemoveAurasDueToSpell(aura->GetId(), aura->GetCasterGUID());
+                    return;
```
Comment (heyitsbench) [inline src/server/scripts/Spells/spell_item.cpp:4132]:
```suggestion
                player->RemoveAurasDueToSpell(aura->GetId(), aura->GetCasterGUID());
                return;
```
Spacing issue here it looks like.

Comment (avarishd) [reply src/server/scripts/Spells/spell_item.cpp]: CI also caught that, nice.

**Review by elthehablo — APPROVED** [review]

**Pull request merged**